### PR TITLE
Auto generate enums for v201

### LINKED ISSUE #40: Auto Generate Enums Class based on json schemas
The idea is to follow the already implemented generator for the call and call result payloads and generate classes for all the OCPP spec enums. After a quick research, I found that there are schemas like the Firmware Status Notification: [https://github.com/mobilityhouse/ocpp/blob/e30a0db5a8afaad24ff9612bad0208381a3af2de/ocpp/v20/schemas/FirmwareStatusNotificationRequest_v1p0.json#L7](https://github.com/mobilityhouse/ocpp/blob/e30a0db5a8afaad24ff9612bad0208381a3af2de/ocpp/v20/schemas/FirmwareStatusNotificationRequest_v1p0.json#L7)

that does not include a definition of the Enumeration and just adds the enums to the property itself which is not the way used in other schemas like in GetVariablesRequest [https://github.com/mobilityhouse/ocpp/blob/e30a0db5a8afaad24ff9612bad0208381a3af2de/ocpp/v20/schemas/GetVariablesRequest_v1p0.json#L4](https://github.com/mobilityhouse/ocpp/blob/e30a0db5a8afaad24ff9612bad0208381a3af2de/ocpp/v20/schemas/GetVariablesRequest_v1p0.json#L4
)

First, we can build a script to generate Enums based on the schemas like the GetVariablesRequest one and see which ones were generated and if there are Enums missing

### PR DESCRIPTION
**Pull request by tropxy**
fixes #40

### CODE REVIEW

File: scripts/schema_to_enums_v201.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+import sys
+import re
+import json
+from pathlib import Path
+
+
+def create_class(name):
+    return normalclass(name)
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:9]: I'm not sure why this function exists. Why not call the constructor of `normalclass` directly?

File: scripts/schema_to_enums_v201.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def create_attribute(name, casing_style="snake_case"):
+    # Assumes the variable name is a concatenation of capitalized therms, e.g.,
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:13]: s/therms/terms

File: scripts/schema_to_enums_v201.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    if casing_style == "CamelCase":
+        name_converted = name_normalized[0].lower() + name_normalized[1:]
+    else:
+        name_converted = re.sub('([a-z0-9])([A-Z])', r'\1_\2',
+                                name_normalized).lower()
+
+    return attribute(name, name_converted)
+
+
+class normalclass:
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:26]: s/normalclass/NormalClass

File: scripts/schema_to_enums_v201.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        if len(self.attrs) == 0:
+            return output + "    pass\n"
+
+        for attr in self.attrs:
+            output += str(attr)
+
+        return output
+
+
+class attribute:
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:46]: s/attribute/Attribute

File: scripts/schema_to_enums_v201.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        self.name_converted = name_converted
+
+    def __str__(self):
+        name = self.name
+        if not re.match("^[a-zA-Z_]", self.name):
+            name = "_" + self.name
+
+        definition = f'    {self.name_converted} = "{name}"\n'
+
+        return definition
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:58]: This can shortened to a single line.

File: scripts/schema_to_enums_v201.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                enum_types_names.append(nc.name)
+
+
+if __name__ == '__main__':
+    if len(sys.argv) != 2:
+        print("Pass path to folder with schemas")
+        sys.exit(-1)
+
+    p = Path(sys.argv[1])
+    schemas = list(p.glob("*.json"))
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:96]: Can you document the arguments?

File: scripts/schema_to_enums_v201.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    for schema in schemas:
+        parse_schema(schema)
+
+    with open('enums.py', 'wb+') as f:
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:101]: IIRC `+` will open the file in a mode where `write()` appends to the file instead of overwriting the original source. So if you run this script twice the output has all enum definitions twice. I'm not sure if you want that.

**Review by OrangeTux — CHANGES_REQUESTED** [review]
Can you add a bit more documentation? A script like won't be used often. So if we've to use or modify it in the future it's important that documentation explains how the code works.

File: scripts/schema_to_enums_v201.py (hunk 1 above)
Comment (tropxy) [reply scripts/schema_to_enums_v201.py:9]: Legacy code from the other script

File: scripts/schema_to_enums_v201.py (hunk 7 above)
Comment (tropxy) [reply scripts/schema_to_enums_v201.py:101]: No, that would be the case if flag 'a' would be used... 'wb+' writes as binary and truncate the file

File: ocpp/charge_point.py [hunk 8]
```diff
@@ -261,7 +261,7 @@ async def call(self, payload):
 
         if response.message_type_id == MessageType.CallError:
             LOGGER.warning("Received a CALLError: %s'", response)
-            return
+            return response
```
Comment (OrangeTux) [inline ocpp/charge_point.py]: This shouldn't be part of this PR. It's already worked on in a different PR>

File: scripts/schema_to_enums_v201.py [hunk 9]
```diff
@@ -92,15 +91,21 @@ def parse_schema(schema):
...
     schemas = list(p.glob("*.json"))
 
-    for schema in schemas:
-        parse_schema(schema)
+    # This loop will update the lists enum_types and enum_types_names:
+    # The former contains the enum objects that are later parsed to the
+    # enum.py, using the str representation. And the latter, is used to keep
+    # track of the already processed Enums, avoiding data duplication
+    for schema_ in schemas:
+        parse_schema(schema_)
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:103]: Can't you use `schema` over `schema_`?

File: scripts/schema_to_enums_v201.py [hunk 10]
```diff
@@ -4,26 +4,29 @@
...
 from pathlib import Path
 
-
-def create_class(name):
-    return normalclass(name)
+enum_types = []
+enum_types_names = []
 
 
 def create_attribute(name, casing_style="snake_case"):
```
Comment (OrangeTux) [inline scripts/schema_to_enums_v201.py:11]: The value of `snake_case` is never used. Wouldn't a boolean more appropriate?

**Review by OrangeTux — CHANGES_REQUESTED** [review]

File: scripts/schema_to_enums_v201.py (hunk 10 above)
Comment (tropxy) [reply scripts/schema_to_enums_v201.py:11]: I was layering ground for more possible case styles. You prefer me to do that change and make it bool? or include it in a if/elif  and even raise an error in case no answer is provided?

File: scripts/schema_to_enums_v201.py (hunk 9 above)
Comment (tropxy) [reply scripts/schema_to_enums_v201.py:103]: My linter complains because schema shadows the use of schema in parse_schema

Comment (OrangeTux) [reply scripts/schema_to_enums_v201.py:103]: Isn't that a problem with your linter? The scopes of both variables are totally different. `schema_` lives inside the main, while `schema` lives inside `parse_schema()`

Comment (tropxy) [reply scripts/schema_to_enums_v201.py]: You are right

File: scripts/schema_to_enums_v201.py (hunk 10 above)
Comment (tropxy) [reply scripts/schema_to_enums_v201.py:11]: Let me know asap

Comment (OrangeTux) [reply scripts/schema_to_enums_v201.py:11]: > I was layering ground for more possible case styles. 

When do you expect changes to output casing? The casing of Python code in our library is static, either snake_case or CamelCase. I don't see that changing soon.

Comment (tropxy) [reply scripts/schema_to_enums_v201.py:11]: Ok, agreed.

**Review by OrangeTux — APPROVED** [review]
